### ISSUE TITLE
PAMR: RMI Dynamic Code downloading does not work

### ISSUE THREAD

**Issue by activeeon-bot** [ow2-proactive/programming#564]
<a href="https://jira.activeeon.com/browse/PROACTIVE-683" title="PROACTIVE-683">Original issue</a> created by <a href="mailto:clement.mathieu_AT_activeeon.com">Clement Mathieu</a> on 28, May 2009 at 17:16 PM - PROACTIVE-683

<hr />

When PAMR is activated, RMI dynamic code downloading (codebase) does not work. The codebase contains an HTTP URL, so RMI performs an HTTP GET to download missing classes. The HTTP GET will most likely fails due to NAT or Firewalls.

The following code exhibits the issue:
    ao = (AO) PAActiveObject.newActive(AO.class.getName(), new Object(link: ) {}, node);
    PAActiveObject.newActive(AO2.class.getName(), new Object(link: ) {ao}, node);

Output with proactive.classloading in debug mode:
  66077@britany.activeeon.com - (link: INFO        proactive.mop) Generating class : pa.stub.org.objectweb.proactive.examples.teststub._StubAO
  66077@britany.activeeon.com - (link: INFO        proactive.mop) Generating class : pa.stub.org.objectweb.proactive.examples.teststub._StubAO2
  66077@britany.activeeon.com - (link: INFO                job.0) 63505@britany.activeeon.com - (link: WARN oactive.classloading) Serving request from /pa/stub/org/objectweb/proactive/examples/teststub/_StubAO.class
  66077@britany.activeeon.com - (link: INFO                job.0) 63505@britany.activeeon.com - (link: INFO        proactive.mop) Generating class : pa.stub.org.objectweb.proactive.examples.teststub._StubAO
  66077@britany.activeeon.com - (link: INFO                job.0) 63505@britany.activeeon.com - (link: DEBUG oactive.classloading) Served request from 192.168.1.22:47463 for pa.stub.org.objectweb.proactive.examples.teststub._StubAO status: OK

As we can see, the stub has been generated by 66077. When AO2 is created 63505 tries to download the stub from 66077.

We must find a way to replace the HTTP GET request by a local stub generation or by registering a new URL protocol handler. This handler will be in charge of downloading the class with a ProActive methodcall instead of an HTTP Get.

As a temporary workaround, all the ProActive stubs can be generated at compile time (by using the StubGenerator) and then put into the application classpath. It the stub is locally available, RMI will not download it from a remote host.

**Comment by activeeon-bot** [comment]
Attachments:
- [pamr_cl_singleton.patch](https://github.com/ow2-proactive/backup-attachments-jira/blob/gh-pages/proactive/683/pamr_cl_singleton.patch)
- [pamr_cl.patch](https://github.com/ow2-proactive/backup-attachments-jira/blob/gh-pages/proactive/683/pamr_cl.patch)

**Comment by activeeon-bot** [comment]
<a href="https://jira.activeeon.com/browse/PROACTIVE-683?focusedCommentId=11587">Original comment</a> posted by <a href="mailto:clement.mathieu_AT_activeeon.com">Clement Mathieu</a> on 04, Jun 2009 at 16:35 PM

<hr />

Raised priority since we need this feature for the new CPER cluster

**Comment by activeeon-bot** [comment]
<a href="https://jira.activeeon.com/browse/PROACTIVE-683?focusedCommentId=11588">Original comment</a> posted by <a href="mailto:clement.mathieu_AT_activeeon.com">Clement Mathieu</a> on 04, Jun 2009 at 16:36 PM

<hr />

The ProActiveRMIClassloader could be used to sovled this issue but it is currently broken

**Comment by activeeon-bot** [comment]
<a href="https://jira.activeeon.com/browse/PROACTIVE-683?focusedCommentId=11600">Original comment</a> posted by <a href="mailto:clement.mathieu_AT_activeeon.com">Clement Mathieu</a> on 05, Jun 2009 at 17:26 PM

<hr />

Status update:

I have made a quick & dirty prototype of dynamic class downloading through ProActive. It barely works. I plan to provide a clean implementation by next week.

Here is how it works:
1- Define a custom RMIClassloader. It delegates all the method call to the standard RMIClassloader but getClassAnnotation(). This method  adds the URL of the local ProActive runtime to the codebase. We cannot use java.rmi.codebase because this property is read only once, so we cannot update it when the runtime becomes ready. 
2- Register custom protocol handlers (parmi, paibis, papamr, pahttp). In step one, the URL of the ProActive runtime is prefixed with "pa". Theses protocols allow to download classes from remote ProActive runtime.

**Comment by activeeon-bot** [comment]
<a href="https://jira.activeeon.com/browse/PROACTIVE-683?focusedCommentId=11612">Original comment</a> posted by <a href="mailto:clement.mathieu_AT_activeeon.com">Clement Mathieu</a> on 12, Jun 2009 at 11:47 AM

<hr />

Patch is available in the trunk, should work !

Jean-luc/christian can you build the Scheduler RCP client with a ProActive.jar from trunk ? I have not been able to do it by myself. Thanks

**Comment by activeeon-bot** [comment]
<a href="https://jira.activeeon.com/browse/PROACTIVE-683?focusedCommentId=11613">Original comment</a> posted by <a href="mailto:clement.mathieu_AT_activeeon.com">Clement Mathieu</a> on 12, Jun 2009 at 15:05 PM

<hr />

Since the fix for this issue is quite invasive (new RMI Classloader !), it will not be merged into the 4.1.x release maintenance branch.
Will be available in 4.2.0.

The StubGenerator must be used in 4.1.x and some application patterns will not work with PAMR in 4.1.x (for example a client registering a callback AO on a server if the server does not have the class  of the AO locally).

**Comment by activeeon-bot** [comment]
<a href="https://jira.activeeon.com/browse/PROACTIVE-683?focusedCommentId=11931">Original comment</a> posted by <a href="mailto:clement.mathieu_AT_activeeon.com">Clement Mathieu</a> on 22, Oct 2009 at 09:50 AM

<hr />

The ProActiveClassloader is available in 4.2.x but disabled by default because it is incompatible with the Eclipse RCP framework.

**Comment by activeeon-bot** [comment]
<a href="https://jira.activeeon.com/browse/PROACTIVE-683?focusedCommentId=12062">Original comment</a> posted by <a href="mailto:florin-alexandru.bratu_AT_sophia.inria.fr">Florin Bratu</a> on 04, Dec 2009 at 19:00 PM

<hr />

The current solution relies on the RMI ClassLoader and the RMI custom serialization process, which is customized in order to use ProActive communications. Just like mentioned in the previous comments, the problem with the current solution is that it is too invasive, and thus incompatible with other existing ProActive components.

Instead, we could consider a different approach:
-  create a custom serialization process, by providing custom implementations for Object(link: Output text: Input)Stream - in the same manner that RMI does. The annotations will be the URLs of the ProActive Runtimes which provide the class definitions
-  provide a custom classloader, which should download the class definitions from the remote runtimes located at the URL read from the annotation
-  create a custom pamr marshaller, which should orchestrate the marshalling/unmarshalling of pamr messages

Although the above-mentioned entities provide (more or less) the same functionality as the RMI serialization process and the RMI classloader, they provide the following benefits:
-  it works if the RMI classloader is unavailable. The availability of the RMI classloader depends on other configuration parameters which are sometimes beyond our control - for instance, the availability of a SecurityManager
-  it reduces the coupling of the message routing implementation code with the Core code and with RMI. The previous solution(revision r13041) proposes a new RMIClassLoader, which is activated irrespective of whether the client uses pamr or another communication protocol. Ideally, the solution to this issue should be activated only if the user uses the message routing protocol.

Please find attached a patch which contains an attempt to solve this issue, according to the above-mentioned guidelines. This patch provides a working solution for the issues encountered within the CPER cluster use case. The patch is made against the ProActive trunk SVN revision 15006.

DISCLAIMER: This patch offers a proof-of-concept implementation, which is for illustration purposes only. It is **not** meant to provide a final solution to this issue. Although it has been tested in some particular scenarios, further (more thorough) testing is required in order to ensure it properly interacts with all other ProActive components. Also, the code itself could use some extra polishing.

**Comment by activeeon-bot** [comment]
<a href="https://jira.activeeon.com/browse/PROACTIVE-683?focusedCommentId=12092">Original comment</a> posted by <a href="mailto:clement.mathieu_AT_activeeon.com">Clement Mathieu</a> on 09, Dec 2009 at 11:02 AM

<hr />

Interesting approach. It is lightweight, and enable the same behaviour with PARM than with RMI.

Please note that the current ProActiveClassLoader does more than only allows RMI like dynamic code loading. For example you can call PAActiveObject.newActive(Yourclass) and the classloader will automatically download the class from the parent runtime (useful with GCM Deployment). That's why it leaves in the core and not in the message routing package. Perhaps we can mix both your and my approach.

Unfortunately I don't have the time to work on this issue now. Please synchronize with arnaud & aurnaud gastinel to check that your work is not redundant or does not conflict with other ongoing developments. Sometimes we are trying to fix the issue at an upper level than PAMR (for example direct connections VS mutli protocol remote objects).

**Comment by activeeon-bot** [comment]
<a href="https://jira.activeeon.com/browse/PROACTIVE-683?focusedCommentId=12152">Original comment</a> posted by <a href="mailto:clement.mathieu_AT_activeeon.com">Clement Mathieu</a> on 13, Jan 2010 at 15:35 PM

<hr />

Patch applied against the trunk. RMIProActiveClassloader is now deprecated

**Comment by activeeon-bot** [comment]
<a href="https://jira.activeeon.com/browse/PROACTIVE-683?focusedCommentId=12236">Original comment</a> posted by <a href="mailto:alberto.pozzato_AT_insiel.it">Alberto Pozzato</a> on 04, Mar 2010 at 10:16 AM

<hr />

Hi all, I wasn't able to get nbody example works due to network configuration (NAT, Firewall). At the end I get it works using pamr protocol and proactive.classloading.useHTTP=false. 
What still doesn't work with this configuration is IC2D RCP. Is there some workaround that permits IC2D to work correctly?

I'm running ProActive 4.2.0.

Thank you.

P.S. I also had to open a tunnel connection externally by mean of ssh (ssh -L 33647:xxx.xxx.xxx.xxx:33647 -N  root@xxx.xxx.xxx.xxx) to permit at my application to connect to Message Router placed on the network reachable for me only through ssh (port 22). Could I obtain it directly in ProActive throug a better configuration?

**Comment by activeeon-bot** [comment]
<a href="https://jira.activeeon.com/browse/PROACTIVE-683?focusedCommentId=12237">Original comment</a> posted by <a href="mailto:florin-alexandru.bratu_AT_sophia.inria.fr">Florin Bratu</a> on 04, Mar 2010 at 11:14 AM

<hr />

Hello Alberto,

I think this discussion is more appropriate for the ProActive mailing list
http://proactive.inria.fr/index.php?page=mailing_list

PS: If you need help, you should provide more detailed information on what does not work. For instance, what are the errors that IC2D reports?

**Comment by activeeon-bot** [comment]
<a href="https://jira.activeeon.com/browse/PROACTIVE-683?focusedCommentId=12372">Original comment</a> posted by <a href="mailto:clement.mathieu_AT_activeeon.com">Clement Mathieu</a> on 27, Apr 2010 at 11:41 AM

<hr />

Afert a quick review the patch cannot work. It will introduce class cast exception if a class is downloaded by two different classloader (one classloader per input stream)

**Comment by activeeon-bot** [comment]
<a href="https://jira.activeeon.com/browse/PROACTIVE-683?focusedCommentId=12382">Original comment</a> posted by <a href="mailto:florin-alexandru.bratu_AT_sophia.inria.fr">Florin Bratu</a> on 28, Apr 2010 at 10:56 AM

<hr />

I've tested the patch, it works fine. Plus it passes the functional tests for messagerouting.

Can you be more specific about the problem you encounter? What is your use case? Can you provide the full stacktrace of the ClassCastException you encounter?

**Comment by activeeon-bot** [comment]
<a href="https://jira.activeeon.com/browse/PROACTIVE-683?focusedCommentId=12386">Original comment</a> posted by <a href="mailto:florin-alexandru.bratu_AT_sophia.inria.fr">Florin Bratu</a> on 28, Apr 2010 at 16:20 PM

<hr />

It is easy to make the PAMR classloader available in the Message Routing ROF and thus make it available per-JVM. Here is a patch which does this(vs trunk r16027)

**Comment by activeeon-bot** [comment]
<a href="https://jira.activeeon.com/browse/PROACTIVE-683?focusedCommentId=13064">Original comment</a> posted by <a href="mailto:clement.mathieu_AT_activeeon.com">Clement Mathieu</a> on 21, Oct 2010 at 11:38 AM

<hr />

Seems to work now. ProActiveRemoteClassloader are used both in PAMR and PNP
